e220/e220.go

```diff
@@ -0,0 +1,349 @@
+// Datasheet(LLCC68): https://semtech.my.salesforce.com/sfc/p/#E0000000JelG/a/2R000000HTJR/Tem0gUxGfOZ2Qn3bUzmV2zKNQRYJ3bpobPfOQ7B.erE
+//
+// LoraWAN 1.1 Specification: https://hz137b.p3cdn1.secureserver.net/wp-content/uploads/2020/11/lorawantm_specification_-v1.1.pdf?time=1673563697
+package e220
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"machine"
+	"runtime"
+	"strings"
+	"time"
+)
+
+// targetUARTbaud represents the UART baud rate of the target board
+const (
+	TargetUARTBaud1200kbps   = 1200
+	TargetUARTBaud2400kbps   = 2400
+	TargetUARTBaud4800kbps   = 4800
+	TargetUARTBaud9600kbps   = 9600
+	TargetUARTBaud19200kbps  = 19200
+	TargetUARTBaud38400kbps  = 38400
+	TargetUARTBaud57600kbps  = 57600
+	TargetUARTBaud115200kbps = 115200
+)
+
+// Mode represents the operating mode of the E220
+const (
+	Mode0 uint8 = iota
+	Mode1
+	Mode2
+	Mode3
+)
+
+// Device wraps E220 LoRa module
+type Device struct {
+	uart               *machine.UART
+	m0                 machine.Pin
+	m1                 machine.Pin
+	tx                 machine.Pin
+	rx                 machine.Pin
+	aux                machine.Pin
+	targetUARTBaudRate uint32
+
+	parameters []byte
+	payload    []byte
+
+	txAddr   uint16
+	txCh     uint8
+	txMethod uint8
+}
+
+const (
+	writingRegisterSize = 8
+	bufferSize          = 400
+)
+
+// New returns a new E220 driver. Pass UART object and each used machine.Pin
+func New(
+	uart *machine.UART,
+	m0, m1, tx, rx, aux machine.Pin,
+	targetUARTBaudRate uint32,
+) *Device {
+	d := &Device{
+		uart:               uart,
+		m0:                 m0,
+		m1:                 m1,
+		tx:                 tx,
+		rx:                 rx,
+		aux:                aux,
+		targetUARTBaudRate: targetUARTBaudRate,
+		parameters:         make([]byte, writingRegisterSize),
+		payload:            make([]byte, bufferSize),
+		// Set to factory setting
+		txAddr:   0xFFFF,
+		txCh:     15,
+		txMethod: TxMethodTransparent,
+	}
+	return d
+}
+
+// Configure sets up the device for communication
+func (d *Device) Configure(startupE220Mode uint8) error {
+	d.uart.Configure(machine.UARTConfig{BaudRate: d.targetUARTBaudRate, TX: d.tx, RX: d.rx})
+	d.m0.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	d.m1.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	d.aux.Configure(machine.PinConfig{Mode: machine.PinInput})
+
+	err := d.SetMode(startupE220Mode)
+	if err != nil {
+		return err
+	}
+
+	// AUX Pin becomes Low Level for about 30ms after the E220 is started.
+	// therefore, wait until transmission/reception becomes possible.
+	time.Sleep(50 * time.Millisecond)
+
+	return nil
+}
+
+// SetMode sets operation mode of E220
+func (d *Device) SetMode(mode uint8) error {
+	switch mode {
+	case Mode0:
+		d.m1.Low()
+		d.m0.Low()
+	case Mode1:
+		d.m1.Low()
+		d.m0.High()
+	case Mode2:
+		d.m1.High()
+		d.m0.Low()
+	case Mode3:
+		d.m1.High()
+		d.m0.High()
+	default:
+		return fmt.Errorf("Invalid mode: %d", mode)
+	}
```

Comment:
this could be reduced to `d.m0.Set(mode&1 != 0); d.m1.Set((mode>>1)&1 != 0)` and a error check if mode is greater than 3